### PR TITLE
PPO: Proximal Policy Algorithms

### PR DESCRIPTION
**Pull request by AntoineTheb**

### CODE REVIEW

File: deep-learning/_posts/2019-05-09-PPO.md [hunk 1]
```diff
@@ -0,0 +1,70 @@
+---
+layout: review
+title:  "Clipped Proximal Policy Optimization Algorithm"
+tags:   deep-learning
```
Comment (jhlegarreta) [inline deep-learning/_posts/2019-05-09-PPO.md]: May add the `reinforcment-learning` tag?

File: deep-learning/_posts/2019-05-09-PPO.md [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+author: Antoine Théberge
+pdf:    "https://arxiv.org/pdf/1707.06347"
+cite:
+    authors: "John Schulman, Filip Wolski, Prafulla Dhariwal, Alec Radford, Oleg Klimov"
+    title:   "Proximal Policy Optimization Algorithms"
+    venue:   "Arxiv"
```
Comment (jhlegarreta) [inline deep-learning/_posts/2019-05-09-PPO.md]: Add the year?

File: deep-learning/_posts/2019-05-09-PPO.md [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+---
+
+## Introduction
+
+PPO is currently a state of the art algorithm for reinforcement learning that builds on TRPO$$^{[1]}$$ and is said to have the stability of trust-region methods while being much simpler to implement, requiring only a few changes to vanilla policy gradient methods.
+
+The paper has two main contributions: the clipped surrogate objective and multiple epochs for policy updates. 
+
+While the paper also proposes another way of constraining the policy updates to a trust region by adding a KL penalty coefficient, this latter approach is both more complicated to understand and implement and has been shown empirically to produce worst results. It will therefore not be explained here.
```
Comment (jhlegarreta) [inline deep-learning/_posts/2019-05-09-PPO.md]: May be
produce worst results : produce worse results
?

File: deep-learning/_posts/2019-05-09-PPO.md (hunk 2 above)
Comment (jodoin) [reply deep-learning/_posts/2019-05-09-PPO.md]: maybe add the real venue.

File: deep-learning/_posts/2019-05-09-PPO.md [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+pdf:    "https://arxiv.org/pdf/1707.06347"
+cite:
+    authors: "John Schulman, Filip Wolski, Prafulla Dhariwal, Alec Radford, Oleg Klimov"
+    title:   "Proximal Policy Optimization Algorithms"
+    venue:   "Arxiv"
+---
+
+## Introduction
+
+PPO is currently a state of the art algorithm for reinforcement learning that builds on TRPO$$^{[1]}$$ and is said to have the stability of trust-region methods while being much simpler to implement, requiring only a few changes to vanilla policy gradient methods.
```
Comment (jodoin) [inline deep-learning/_posts/2019-05-09-PPO.md]: state-of-the-art

File: deep-learning/_posts/2019-05-09-PPO.md [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+While the paper also proposes another way of constraining the policy updates to a trust region by adding a KL penalty coefficient, this latter approach is both more complicated to understand and implement and has been shown empirically to produce worst results. It will therefore not be explained here.
+
+## Clipped Surrogate Objective
+
+Vanilla policy gradient methods work by optimizing the following loss
+
+![](/deep-learning/images/PPO/pg.jpg)
+
+where $$\hat{A}$$ is the advantage function. By performing gradient ascent on this, you will take actions that lead to higher rewards more likely. With the idea of importance sampling, we can replace the objective function with $$\hat{r}$$ =
```
Comment (jodoin) [inline deep-learning/_posts/2019-05-09-PPO.md]: $$\hat{r}$$ = 

is this a typo?

File: deep-learning/_posts/2019-05-09-PPO.md [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+![](/deep-learning/images/PPO/surrogate.jpg)
+
+The KL term is to bound the improvement step to a trust region.
+
+But using a KL divergence term is complicated to implement and hard to understand. Instead, the authors decided to bound the policy updated with the following Clipped Surrogate Objective function:
+
+![](/deep-learning/images/PPO/clipped.jpg)
+
+Here, $$\hat{r}$$ is the previously mentioned policy ratio and epsilon is a hyperparameter (0.2 seems to work well). The clip function prevents the ratio going away from the interval $$(1 - \epsilon, 1 + \epsilon)$$. Taking the minimum of the clipped and unclipped objective gives a lower bound on the final objective function.
```
Comment (jodoin) [inline deep-learning/_posts/2019-05-09-PPO.md]: Here, $$r(\theta)$$

File: deep-learning/_posts/2019-05-09-PPO.md (hunk 2 above)
Comment (AntoineTheb) [reply deep-learning/_posts/2019-05-09-PPO.md]: I can't seem to find where it was published, everything points to Arxiv.

File: deep-learning/_posts/2019-05-09-PPO.md (hunk 5 above)
Comment (AntoineTheb) [reply deep-learning/_posts/2019-05-09-PPO.md]: No, it's to add a hat to the variable "r"

File: deep-learning/_posts/2019-05-09-PPO.md (hunk 2 above)
Comment (jodoin) [reply deep-learning/_posts/2019-05-09-PPO.md]: ok

File: deep-learning/_posts/2019-05-09-PPO.md (hunk 1 above)
Comment (jhlegarreta) [reply deep-learning/_posts/2019-05-09-PPO.md]: Done.

**Pull request merged**